### ISSUE TITLE
Add ability for aws_glue_job to use Python 3 and glue version 1.0

### ISSUE THREAD

**Issue by robertaves** [terraform-providers/terraform-provider-aws#9524]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

AWS Glue now supports the ability to run ETL jobs on Apache Spark 2.4.3 (with Python 3).. terraform support needed

### New or Affected Resource(s)

aws_glue_job

### Potential Terraform Configuration

```hcl
resource "aws_glue_job" "aws_glue_job_foo" {
  glue_version = "1"
  name         = "job-name"
  description  = "job-desc"
  role_arn     = data.aws_iam_role.aws_glue_iam_role.arn
  max_capacity = 1
  max_retries  = 1
  connections  = [aws_glue_connection.connection.name]
  timeout      = 5

  command {
    name            = "pythonshell"
    script_location = "s3://bucket/script.py"
    python_version  = "3"
  }

  default_arguments = {    
    "--job-language" = "python"
    "--ENV"          = "env"
    "--ROLE_ARN"     = data.aws_iam_role.aws_glue_iam_role.arn
  }

  execution_property {
    max_concurrent_runs = 1
  }
}
```

### References
* https://aws.amazon.com/about-aws/whats-new/2019/07/aws-glue-now-supports-ability-to-run-etl-jobs-apache-spark-243-with-python-3/
* https://docs.aws.amazon.com/cli/latest/reference/glue/create-job.html

**Comment by ewbankkit** [comment]
It looks like this is enabled via the [Glue version](https://docs.aws.amazon.com/glue/latest/dg/add-job.html#create-job) for a job, added in [AWS SDK v1.21.4](https://github.com/aws/aws-sdk-go/releases/tag/v1.21.4).
Requires:
* https://github.com/terraform-providers/terraform-provider-aws/pull/9489

**Comment by nywilken** [comment]
Related: #9409

**Comment by ezidio** [comment]
+1

**Comment by ezidio** [comment]
Alternative to set python and glue version


```
resource "aws_glue_job" "etl" {
  name     = "${var.job_name}"
  role_arn = "${var.iam_role_arn}"

  command {
    script_location = "s3://${var.bucket_name}/${aws_s3_bucket_object.script.key}"
  }

  default_arguments = {
    "--enable-metrics" = ""
    "--job-language" = "python"
    "--TempDir" = "s3://${var.bucket_name}/TEMP"
  }

  # Manually set python 3 and glue 1.0
  provisioner "local-exec" {
    command = "aws glue update-job --job-name etl_fake --job-update 'Command={ScriptLocation=s3://${var.bucket_name}/${aws_s3_bucket_object.script.key},PythonVersion=3,Name=glueetl},GlueVersion=1.0,Role=${var.iam_role_arn},DefaultArguments={--enable-metrics=\"\",--job-language=python,--TempDir=\"s3://${var.bucket_name}/TEMP\"}'"
  }
}
```

**Comment by Vedant-R** [comment]
Any idea when does this change get pushed?

**Comment by jarosmpost** [comment]
I will do the same as @ezidio suggests, but with job-language scala instead of python.
I also think it would be good if it would work without this workaround.

**Comment by whitney** [comment]
Maybe more urgent given the [announcement](https://www.python.org/doc/sunset-python-2/) of python 2's official sunsetting on Jan 1 2020.

**Comment by ztane** [comment]
the "workaround" is a horrible PITA as all the arguments need to be flattened into one string... with nested escaping.

**Comment by ztane** [comment]
In fact there is **no proper workaround** because any modification will reset the job back to Python 2. The local-exec provisioner will not be rerun.

**Comment by Vedant-R** [comment]
> In fact there is **no proper workaround** because any modification will reset the job back to Python 2. The local-exec provisioner will not be rerun.

I think the below script logic should do the job for you. Using null resource based on timestamp.

```
resource "aws_glue_job" "etl" {
  name     = "${local.name}"
  role_arn = "${module.crawler_role.role_arn}"

  command {
    script_location = "s3://abc/abc.py"
  }

  default_arguments = {
    "--job-language" = "python"
    "--database"     = "${local.name}"
    "--s3bucket"    = "${var.bucket_name}"
  }
}

resource "null_resource" "cluster" {
  depends_on = ["aws_glue_job.etl"]

  triggers = {
    time = "${timestamp()}"
  }

  provisioner "local-exec" {
    command = "aws glue update-job --job-name ${local.name} --job-update 'Role=${module.crawler_role.role_arn}, Command={ScriptLocation=s3://abc/abc.py,PythonVersion=3,Name=glueetl}, DefaultArguments={--job-language=python,--database=${local.name},--s3bucket=<bucket-name>}, Connections={Connections=[${local.name}]}, GlueVersion=1.0'"
  }
}
```

**Comment by g-sree** [comment]
The python_version  = "3" option is enabled in latest provider `terraform-provider-aws_v2.29.0`, however that did not modify "Spark version" on an existing job, because of which the job failed with the following error.

`JobName:XXXXXXX and JobRunId:jr_XXXXXXXXX failed to execute with exception Unsupported pythonVersion 3 for given glueVersion 0.9`

**Comment by vickymca2005** [comment]
> The python_version = "3" option is enabled in latest provider `terraform-provider-aws_v2.29.0`, however that did not modify "Spark version" on an existing job, because of which the job failed with the following error.
> 
> `JobName:XXXXXXX and JobRunId:jr_XXXXXXXXX failed to execute with exception Unsupported pythonVersion 3 for given glueVersion 0.9`

I am also getting the same issue. What i am missing?

**Comment by g-sree** [comment]
> > The python_version = "3" option is enabled in latest provider `terraform-provider-aws_v2.29.0`, however that did not modify "Spark version" on an existing job, because of which the job failed with the following error.
> > `JobName:XXXXXXX and JobRunId:jr_XXXXXXXXX failed to execute with exception Unsupported pythonVersion 3 for given glueVersion 0.9`
> 
> I am also getting the same issue. What i am missing?

For time being I have deployed with python_version 3 and then from AWS console modified the job with glueVersion 1.. This fixed it.. However its good to have a fix from the provider

**Comment by RamaIndia** [comment]
This issue can be fixed through Cloud formation template. In cloud formation template we can directly declare Glue version and Python version. It will be very easy way and no need update aws provider.

{
	"Description": "AWS Glue Job ",
	"Resources": {
		"GlueJob": {
			  "Type": "AWS::Glue::Job",
			  "Properties": {
					"Command": {
					  "Name": "glueetl",
					  "ScriptLocation": "${script_location}",
					  "PythonVersion" : "3"
					},
					"DefaultArguments": {
					  "--job-language": "${job-language}",
					  "--TempDir" : "${TempDir}",
					  "--extra-jars" : "${extra-jars}"
					},
					"Name": "${Name}",
					"Role": "${role_arn}",
					"MaxCapacity" : 10,
					"GlueVersion" : "1.0"
				}
		}
	}
}

**Comment by ztane** [comment]
@g-sree **all** updates using Terraform will always reset the Python version...

**Comment by g-sree** [comment]
> updates using Terraform will always reset the Python version...

This is my terraform declaration .. 

```
resource "aws_glue_job" "test_glue_job" {
  name     = "name"
  role_arn = "iam_role"

  command {
    script_location = "script"
    python_version  = 3
  }

  default_arguments = {
    ~~~~ truncated ~~~~
  }
}
```

I'm using python3.. however if you modify an existing job only python version changes and not the glue version which should be 1.0 . In this case the job will fail in the next run. I have manually updated the glue version from 0.9 to 1.0 in the AWS console. I never had a problem afterwards .

**Comment by mmell** [comment]
A cloudformation-based workaround: https://github.com/terraform-providers/terraform-provider-aws/issues/8526#issuecomment-490161140

```
resource "aws_cloudformation_stack" "network" {
  name = "${local.name}-glue-job"

  template_body = <<STACK
{
  "Resources" : {
    "MyJob": {
      "Type": "AWS::Glue::Job",
      "Properties": {
        "Command": {
          "Name": "glueetl",
          "ScriptLocation": "s3://${local.bucket_name}/jobs/${var.job}"
        },
        "ExecutionProperty": {
         "MaxConcurrentRuns": 2
        },
        "MaxRetries": 0,
        "Name": "${local.name}",
        "Role": "${var.role}"
      }
    }
  }
}
STACK
}
```

**Comment by bflad** [comment]
Support for the new `glue_version` argument in the `aws_glue_job` resource has been merged and will release with version 2.34.0 of the Terraform AWS Provider, on Thursday. 👍

**Comment by bflad** [comment]
This has been released in [version 2.34.0 of the Terraform AWS provider](https://github.com/terraform-providers/terraform-provider-aws/blob/master/CHANGELOG.md#2340-october-31-2019). Please see the [Terraform documentation on provider versioning](https://www.terraform.io/docs/configuration/providers.html#provider-versions) or reach out if you need any assistance upgrading.

**Comment by fredde975** [comment]
> This has been released in [version 2.34.0 of the Terraform AWS provider](https://github.com/terraform-providers/terraform-provider-aws/blob/master/CHANGELOG.md#2340-october-31-2019). Please see the [Terraform documentation on provider versioning](https://www.terraform.io/docs/configuration/providers.html#provider-versions) or reach out if you need any assistance upgrading.

Hi! 
I'm in need up upgrading the AWS provider. Currently my config looks like this:

- Terraform v0.12.17
-  provider.aws v2.28.1

I have installed the latest Terraform with brew and I tried updating the AWS provider with "terraform init -upgrade". I should in theory get version 2.41.0 (December 04, 2019) but I'm stuck with v2.28.1. 

Any help in this matter would be greatly appreciated.
